### ISSUE TITLE
thread safety

### ISSUE THREAD

**Issue by galderic** [lemire/RoaringBitmap#35]
Hi,

Have you considered making roaring bitmaps thread safe so they can be read while being updated? With read-write reentrant locks at container level the thread contention would be very low. 

I made some tests with external synchronization and the performance penalty is already very low by locking on the whole segment.

Would such contribution be welcomed and if so what would be the best approach? A child implementation synchronizing on protected methods maybe ?

**Comment by lemire** [comment]
@galderic 

Could there be a way to implement this as an optimal API, without penalizing current methods?

**Comment by owenkaser** [comment]
Did you mean 'optional' instead of 'optimal'?

**Comment by pwebbitrs** [comment]
+1 For making this an optional change. Personally, I would be somewhat adverse to any additional penalties on the current implementation. So if this can be done as an optional ThreadSafe version/API (separate to the current version), that would strongly be my preference.   

Also, please correct me if I'm wrong, ( I haven't thought this through completely!), but I'm thinking a thread safe API may *need* to be explicit (as in ThreadSafeRoaringBitmap ) because it would conflict with the mutable/immutable semantics of memory mapped bitmaps?

**Comment by lemire** [comment]
@owenkaser  I did. ;-)

**Comment by lemire** [comment]
@pwebbitrs 

I think that ImmutableRoaringBitmap are thread-safe by default, being immutable.

Otherwise, I think it is prudent to avoid adding overhead that everyone must pay when it might benefit only some users. Thus, if it is possible to add new functions calls instead of modifying the existing ones, it might be better.

**Comment by galderic** [comment]
Nice to see we're all on the same page. In fact @pwebbitrs my approach has been exactly making a ThreadSafeRoaringBitmap extend RoaringBitmap and add support for parallel reads and updates. I only lacked some more visibility on the structural modifications of the highLowContainer array.

I'll try to get my company approval to publish the changes ( it was done during work hours ) so we can continue the discussion. Thanks @lemire @owenkaser  and @pwebbitrs .

**Comment by lemire** [comment]
@galderic  Looking forward to see your work.

**Comment by lemire** [comment]
@galderic  Any luck on this front?

**Comment by galderic** [comment]
@lemire I already have the approval from my company to publish the changes. Current status is that I can check, remove and add bits concurrently by synchronizing in the child class, without much performance penalty. I still have to figure out how to iterate also. I hope I'll be able to commit something near soon.

**Comment by lemire** [comment]
@galderic  Please consider checking the latest release (0.5.1) as we have added a "runOptimize" function that will improve compression in some cases.

**Comment by gpunti** [comment]
@lemire Ok, no problem. It can be called when holding the write lock for that container. I've committed a first version of ConcurrentBitmap in https://github.com/gpunti/RoaringBitmap/tree/thread-safety.  There's a detailed description of what's been done in the wiki page. I'll be happy to read your opinions.

**Comment by lemire** [comment]
@gpunti 

I am currently traveling but I will review this when I come back.

**Comment by lemire** [comment]
@gpunti  Sorry for the delay. I think that your approach makes a whole lot of sense.

It would make a great PR.

Two things...

1. You probably want to sync with the 0.5 release, if you have not done so already.

2. If not done already, make sure to add some unit tests.

**Comment by gpunti** [comment]
Hi @lemire , yes it's sync with the 0.5 release. There're also unit tests checking that maps perform well under heavy parallelism.

What I'd like to know are the forum's thoughts about how to integrate ConcurrentBitmap with the rest of classes. Should we rewrite and all methods ( serialize, and, or, xor, flip,...)  or just add a method to convert it to a regular RoaringBitmap? Wouldn't three version be a little too much? ( roaring, buffer, concurrent)

**Comment by lemire** [comment]
@gpunti 

It is indeed a software engineering issue.

There are a few distinct strategies...

  *  Add a few methods to existing classes...
  *  Wrap existing classes...
  *  Use inheritance...

You have opted to create an altogether new class which adds a maintenance burden or limits flexibility.

Why did you choose to create a new class?

**Comment by gpunti** [comment]
It's explained in detail in https://github.com/gpunti/RoaringBitmap/wiki.  Extending or using delegation didn't seem a good idea to me because:
- (a) Synchronizing highLowContainer externally was very expensive
- (b) Inheritance's is a 'is-a' relationship, so a ConcurrentBitmap and a RoaringBitmap should be interoperable with 'and','or','xor', etc...but RoaringBitmap relies on the RoaringArray structure which ConcurrentBitmap doesn't use because of (a). Also the methods are static and cannot be overriden on runtime.

To make the three branches compatible, I think that all operations ( 'and', 'or', 'serialize', ...) should be on different classes (as in FastAggregation) and agnostic on how the containers are stored or retrieved So the highLowContainer could be an interface with only four operations like

 - contains(short hb) : boolean
 - addContainer(short hb, Container container) : void
 - removeContainer(short hb) : void
 - iterator() : Iterator\<Container\>

In short, have the same abstraction level in the highLowContainer than we have in ImmutableBitmapDataProvider for the containers.

In this way we could have as many different implementation (mapped, thread-safe, ...) as we desired because all of them would be interoperable and any operator ( 'and', 'or', ...) implementation would make no assumptions on how containers are stored.

**Comment by lemire** [comment]
> Extending or using delegation didn't seem a good idea

Seems reasonable.

> In short, have the same abstraction level in the highLowContainer than we have in ImmutableBitmapDataProvider for the containers.

We have this level of abstraction already, at least in the buffer package 


https://github.com/lemire/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/buffer/PointableRoaringArray.java

If you look at the code in ImmutableRoaringBitmap, it does not "know" how highLowContainer is implemented, it just calls it through a PointableRoaringArray interface.

We could generalize this approach throughout the code.

**Comment by gpunti** [comment]
That's true on the or() method but in the and() method it makes use of getKeyAtIndex() and advanceUntil() which relies on it being an array. Both of them also construct a MutableRoaringBitmap by getting its getMappeableRoaringArray() again relying on the result being stored in an array of containers.

**Comment by lemire** [comment]
@gpunti 

Right. I do not think that these methods are specific to an array however. You could get the same effect with 

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentSkipListMap.html


Which brings another issue. Because you use a ConcurrentHashMap, you do not iterate through the contained values in order. If you were to use a ConcurrentSkipListMap, you would recover a behaviour closer to the original classes (which iterate through the values in sorted order).

Thoughts?

**Comment by gpunti** [comment]
That's a good example. ConcurrentSkipListMap implements a Map. Internally it might use a list also ( as its name suggests) but you cannot access it by index position, only by key or value as any other map. 

Regarding the order of elements, it's an interesting question. In fact it does iterate over the values in sorted order because it sorts the list on every container-switch. That's expensive on the iterate() operation, but faster in the add() remove() operations.  I made the assumption that add() and remove() were more important, but actually might not ( I don't know how roaring bitmaps are used by other projects/companies )

I'm trying to figure out how we could operate with bitmaps (and,or,...) regardless of how they are stored. If I found something interesting I'll commit it in my branch for your consideration.

**Comment by lemire** [comment]
@gpunti 

> That's a good example. ConcurrentSkipListMap implements a Map. Internally it might use a list also ( as its name suggests) but you cannot access it by index position, only by key or value as any other map.

Right, one would need something different from the PointableRoaringArray.java, more general. I was not implying that it was usable 'as is'.

> That's expensive on the iterate() operation, but faster in the add() remove() operations. I made the assumption that add() and remove() were more important, but actually might not ( I don't know how roaring bitmaps are used by other projects/companies )


Typically, one uses a compressed bitmap for data that is queried far more than it is updated. So having a fast "iterate()" is important, I would think.

**Comment by JasonHuangHuster** [comment]
@gpunti  i didn't find the `ConcurrentBitMap` in the latest realease which is (0.6) , what's the plan ?

**Comment by gpunti** [comment]
It's implemented and working in a branch. It's not merged in master as I didn't want to introduce a third type of bitmap ( with its own and,or,xor, etc... operations ). If anyone else is interested in concurrent bitmaps, maybe we could revive the discussion on how to unify them in a single  bitmap type, and a single boolean operations set.

**Comment by maxdemarzi** [comment]
@gpunti Looks like your last commit on the forked branch is broken. You may have forgotten to check in the org.roaringbitmap.concurrent.BitmapInterface file.  Actually ran into a use case where this would have come in handy.

**Comment by gpunti** [comment]
Hi @maxdemarzi you are right. I just committed the missing files.

**Comment by antal-kajzer-epam** [comment]
hi guys,
I'm interested whats the status of this topic? I'm getting similar issue like https://github.com/RoaringBitmap/RoaringBitmap/issues/402, but after few days of checking I was not able to catch the thread-safe issue here. Do you know if there will be ThreadSafeRoaringBitMap released some time?
